File: chanclol/proxy.py
import logging
import aiohttp
from enum import Enum

import rate_limiter

logger = logging.getLogger(__name__)


# Responses provided by the Riot API
class ResponseStatus(Enum):
    OK = 200
    BAD_REQUEST = 400
    UNAUTHORIZED = 401
    FORBIDDEN = 403
    DATA_NOT_FOUND = 404
    METHOD_NOT_ALLOWED = 405
    UNSUPPORTED_MEDIA_TYPE = 415
    RATE_LIMIT_EXCEEDED = 429
    INTERNAL_SERVER_ERROR = 500
    BAD_GATEWAY = 502
    SERVICE_UNAVAILABLE = 503
    GATEWAY_TIMEOUT = 504


messages = {
    ResponseStatus.OK: "OK",
    ResponseStatus.BAD_REQUEST: "Bad request",
    ResponseStatus.UNAUTHORIZED: "Unauthorized",
    ResponseStatus.FORBIDDEN: "Forbidden",
    ResponseStatus.DATA_NOT_FOUND: "Data not found",
    ResponseStatus.METHOD_NOT_ALLOWED: "Method not allowed",
    ResponseStatus.UNSUPPORTED_MEDIA_TYPE: "Unsupported media type",
    ResponseStatus.RATE_LIMIT_EXCEEDED: "Rate limit exceeded",
    ResponseStatus.INTERNAL_SERVER_ERROR: "Internal server error",
    ResponseStatus.BAD_GATEWAY: "Bad gateway",
    ResponseStatus.SERVICE_UNAVAILABLE: "Service unavailable",
    ResponseStatus.GATEWAY_TIMEOUT: "Gateway timeout",
}


class Proxy:
    def __init__(self, api_key, restrictions):
        self.rate_limiter = rate_limiter.RateLimiter(restrictions)
        self.api_key = api_key

    # Make a simple async request
    async def request(self, url, vital):

        # Wait until the request can be made
        allowed = await self.rate_limiter.allowed(vital)
        if not allowed:
            logger.warning("Rate limiter is not allowing the request")
            return None

        # Make the request and check the response status
        logger.debug(f"Making a request to url {url}")
        async with aiohttp.ClientSession() as session:
            async with session.get(url, headers=self._build_api_header()) as r:
                return await self._create_response(r)

    # Return a header that includes the API key
    def _build_api_header(self):
        if self.api_key == None:
            logger.error("API key is not available")
            return None
        return {"X-Riot-Token": self.api_key}

    # Create an internal response with the aiohttp response, logging in the
    # process the relevant status codes for this application
    async def _create_response(self, response):

        if not response:
            return None

        # Log accorging to the status
        message = messages[ResponseStatus(response.status)]
        match ResponseStatus(response.status):
            case ResponseStatus.OK | ResponseStatus.DATA_NOT_FOUND:
                logger.debug(message)
            case ResponseStatus.RATE_LIMIT_EXCEEDED:
                logger.warning(message)
                self.rate_limiter.received_rate_limit()
            case _:
                logger.error(message)

        match ResponseStatus(response.status):
            case ResponseStatus.OK:
                return await response.json()
            case _:
                return None

File: internal/common/proxy.go
package common

import (
	"fmt"
	"io"
	"net/http"
	"strconv"
	"time"

	"github.com/rs/zerolog/log"
)

const (
	OK                     int = 200
	BAD_REQUEST            int = 400
	UNAUTHORIZED           int = 401
	FORBIDDEN              int = 403
	DATA_NOT_FOUND         int = 404
	METHOD_NOT_ALLOWED     int = 405
	UNSUPPORTED_MEDIA_TYPE int = 415
	RATE_LIMIT_EXCEEDED    int = 429
	INTERNAL_SERVER_ERROR  int = 500
	BAD_GATEWAY            int = 502
	SERVICE_UNAVAILABLE    int = 503
	GATEWAY_TIMEOUT        int = 504
)

var messages = map[int]string{
	OK:                     "OK",
	BAD_REQUEST:            "Bad request",
	UNAUTHORIZED:           "Unauthorized",
	FORBIDDEN:              "Forbidden",
	DATA_NOT_FOUND:         "Data not found",
	METHOD_NOT_ALLOWED:     "Method not allowed",
	UNSUPPORTED_MEDIA_TYPE: "Unsupported media type",
	RATE_LIMIT_EXCEEDED:    "Rate limit exceeded",
	INTERNAL_SERVER_ERROR:  "Internal server error",
	BAD_GATEWAY:            "Bad gateway",
	SERVICE_UNAVAILABLE:    "Service unavailable",
	GATEWAY_TIMEOUT:        "Gateway timeout",
}

type Proxy struct {
	header      map[string]string
	client      http.Client
	rateLimiter RateLimiter
}

func NewProxy(header map[string]string, restrictions []Restriction) Proxy {
	return Proxy{header, http.Client{}, NewRateLimiter(restrictions)}
}

// Make a request to the provided url, indicating if it is vital.
// The request will be performed depending on the status of the rate limiter
func (proxy *Proxy) Request(url string, vital bool) []byte {

	// ask for permission to execute the request
	// and wait if necessary
	allowedChan := make(chan bool)
	go proxy.rateLimiter.Allowed(vital, allowedChan)
	allowed := <-allowedChan
	if !allowed {
		log.Warn().Msg("Rate limiter is not allowing the request")
		return nil
	}

	// Create the request and add the header
	request, err := http.NewRequest("GET", url, nil)
	if err != nil {
		log.Error().Msg(fmt.Sprintf("Could not create request for url %s", url))
		return nil
	}
	for key, value := range proxy.header {
		request.Header.Set(key, value)
	}

	// Perform the request
	res, err := proxy.client.Do(request)
	if err != nil {
		log.Error().Msg("Could not perform request")
		return nil
	}
	defer res.Body.Close()

	// Check if the status of the request is understood
	message, ok := messages[res.StatusCode]
	if !ok {
		log.Error().Msg(fmt.Sprintf("Status code of request (%d) is not understood", res.StatusCode))
		return nil
	}

	logMessage := fmt.Sprintf("%d %s", res.StatusCode, message)
	switch res.StatusCode {
	case OK:
		log.Debug().Msg(logMessage)
		// Read the response
		stream, err := io.ReadAll(res.Body)
		if err != nil {
			log.Debug().Msg(fmt.Sprintf("Could not extract the response for url %s", url))
			return nil
		}
		return stream
	case DATA_NOT_FOUND:
		log.Debug().Msg(logMessage)
		return nil
	case RATE_LIMIT_EXCEEDED:
		log.Warn().Msg(logMessage)
		proxy.rateLimiter.ReceivedRateLimit(retryAfter(res.Header))
		return nil
	default:
		log.Error().Msg(logMessage)
		// Read body for more info
		bodyBytes, err := io.ReadAll(res.Body)
		if err != nil {
			log.Error().Err(err).Msg("Could not read response body")
			return nil
		}
		// Print the raw body
		log.Error().Msgf("Body: %s", string(bodyBytes))
		// Print headers (useful for rate limits!)
		for key, val := range res.Header {
			log.Debug().Msgf("Header %s: %v", key, val)
		}
		return nil
	}
}

func retryAfter(header http.Header) time.Duration {
	for _, name := range []string{"Retry-After", "X-Rate-Limit-Reset-After"} {
		value := header.Get(name)
		if value == "" {
			continue
		}
		seconds, err := strconv.ParseFloat(value, 64)
		if err != nil {
			log.Warn().Err(err).Msg(fmt.Sprintf("Could not parse %s header %q", name, value))
			continue
		}
		return time.Duration(seconds*float64(time.Second)) + 250*time.Millisecond
	}

	return 0
}
